<a href="https://colab.research.google.com/github/NikitaGalkinV/Data-Analyst-Yandex-Practicum-/blob/main/SQL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div style="border:solid green 2px; padding: 20px">
    
<b>Никита, привет! Это Александр Матвеевский</b> Я рад тебя видеть на финальном спринте (SQL).

---

Я буду красить комментарии цветом, пожалуйста, не удаляй их:

<div class="alert alert-block alert-danger">✍
    

__Комментарий от тимлида №1__

Такой комментарий нужно исправить обязательно, он критически влияет на удачное выполнение проекта.
</div>
    
---

<div class="alert alert-block alert-warning">📝
    

__Комментарий от тимлида №1__


Такой комментарий является рекомендацией или советом. Можешь использовать их на своё усмотрение.
</div>

---

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Такой комментарий  говорит о том, что было сделано что-то качественное и правильное =)
</div>
    
---
    
Предлагаю работать над проектом в диалоге: если у тебя возникают вопросы — пиши об этом в пачке, не стесняйся. Так же, мне будет легче отследить изменения, если ты выделишь свои комментарии:   
    
<div class="alert alert-info"> <b>Комментарии студента:</b> Например, вот так.</div>
    
    
---

# Проект по SQL


## Описание проекта

Ваша компания купила крупный сервис для чтения книг по подписке. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.


**Цель:** проанализировать базу данных

## Описание данных

**Таблица books**

Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

**Таблица publishers**

Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

![image-2.png](attachment:image-2.png)

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Здорово, что не забываешь про оформление проекта
</div>

## Исследование данных

In [ ]:
import pandas as pd
import sqlalchemy
from sqlalchemy import create_engine 

In [ ]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 


In [ ]:
def get_dataset(query, con):
    engine = sqlalchemy.create_engine(con)
    df = pd.read_sql(query, engine)
    return df

In [ ]:

def load_and_print_dataset(query, name):   
    df = get_dataset(query, connection_string)
    print(name)
    display(df)
    print("–––––––––––––––––––––––––––––––––––––––––––––––––")
    display(df.info())
    print("–––––––––––––––––––––––––––––––––––––––––––––––––")
    display(df.describe(include='all'))



In [ ]:
# загрузка и вывод датасета books
query_books = "SELECT * FROM books"
load_and_print_dataset(query_books, 'Книги')


Книги


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

–––––––––––––––––––––––––––––––––––––––––––––––––
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

–––––––––––––––––––––––––––––––––––––––––––––––––


book_id    author_id                title   num_pages  \
count   1000.000000  1000.000000                 1000  1000.00000   
unique          NaN          NaN                  999         NaN   
top             NaN          NaN  Memoirs of a Geisha         NaN   
freq            NaN          NaN                    2         NaN   
mean     500.500000   320.417000                  NaN   389.11100   
std      288.819436   181.620172                  NaN   229.39014   
min        1.000000     1.000000                  NaN    14.00000   
25%      250.750000   162.750000                  NaN   249.00000   
50%      500.500000   316.500000                  NaN   352.00000   
75%      750.250000   481.000000                  NaN   453.00000   
max     1000.000000   636.000000                  NaN  2690.00000   

       publication_date  publisher_id  
count              1000   1000.000000  
unique              618           NaN  
top          2004-06-01           NaN  
freq                 10           NaN  
mean                NaN    171.270000  
std                 NaN     99.082685  
min                 NaN      1.000000  
25%                 NaN     83.000000  
50%                 NaN    177.500000  
75%                 NaN    258.000000  
max                 NaN    340.000000

In [ ]:

# загрузка и вывод датасета authors
query_authors = "SELECT * FROM authors"
load_and_print_dataset(query_authors, 'Авторы')


Авторы


author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

–––––––––––––––––––––––––––––––––––––––––––––––––
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

–––––––––––––––––––––––––––––––––––––––––––––––––


author_id                                             author
count   636.000000                                                636
unique         NaN                                                636
top            NaN  Umberto Eco/Girolamo De Michele/Alastair McEwe...
freq           NaN                                                  1
mean    318.500000                                                NaN
std     183.741666                                                NaN
min       1.000000                                                NaN
25%     159.750000                                                NaN
50%     318.500000                                                NaN
75%     477.250000                                                NaN
max     636.000000                                                NaN

In [ ]:

# загрузка и вывод датасета publishers
query_publishers = "SELECT * FROM publishers"
load_and_print_dataset(query_publishers, 'Издатели')


Издатели


publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

–––––––––––––––––––––––––––––––––––––––––––––––––
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

–––––––––––––––––––––––––––––––––––––––––––––––––


publisher_id            publisher
count     340.000000                  340
unique           NaN                  340
top              NaN  Random House Anchor
freq             NaN                    1
mean      170.500000                  NaN
std        98.293777                  NaN
min         1.000000                  NaN
25%        85.750000                  NaN
50%       170.500000                  NaN
75%       255.250000                  NaN
max       340.000000                  NaN

In [ ]:

# загрузка и вывод датасета ratings
query_ratings = "SELECT * FROM ratings"
load_and_print_dataset(query_ratings, 'Рейтинг')


Рейтинг


rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

–––––––––––––––––––––––––––––––––––––––––––––––––
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

–––––––––––––––––––––––––––––––––––––––––––––––––


rating_id      book_id username       rating
count   6456.000000  6456.000000     6456  6456.000000
unique          NaN          NaN      160          NaN
top             NaN          NaN   paul88          NaN
freq            NaN          NaN       56          NaN
mean    3228.500000   510.574195      NaN     3.928284
std     1863.831001   284.141636      NaN     0.943303
min        1.000000     1.000000      NaN     1.000000
25%     1614.750000   291.000000      NaN     3.000000
50%     3228.500000   506.000000      NaN     4.000000
75%     4842.250000   750.000000      NaN     5.000000
max     6456.000000  1000.000000      NaN     5.000000

In [ ]:

# загрузка и вывод датасета reviews
query_reviews = "SELECT * FROM reviews"
load_and_print_dataset(query_reviews, 'Отзывы')

Отзывы


review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

–––––––––––––––––––––––––––––––––––––––––––––––––
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

–––––––––––––––––––––––––––––––––––––––––––––––––


review_id      book_id username  \
count   2793.000000  2793.000000     2793   
unique          NaN          NaN      160   
top             NaN          NaN  susan85   
freq            NaN          NaN       29   
mean    1397.000000   504.693161      NaN   
std      806.413976   288.472931      NaN   
min        1.000000     1.000000      NaN   
25%      699.000000   259.000000      NaN   
50%     1397.000000   505.000000      NaN   
75%     2095.000000   753.000000      NaN   
max     2793.000000  1000.000000      NaN   

                                                     text  
count                                                2793  
unique                                               2793  
top     Foot alone blood anyone building. Also confere...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

**Вывод:**

- В данных отсутствуют пропуски
- Всего было выпущено 1000 книг на 636 авторов с средним кол-вом страниц 230 - это достаточно маленькая база
- Всего 340 издательств
- Медиана рейтинга 3.9, в базе довольно хорошие книги (но это не точно)
- Пользователи оставили почти 2800 отзывов, для выбора книг это будет очень полезно

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Отлично первый взгляд на данные есть. Ты корректно выгрузил данные и посмотрел их содержимое
</div>

### Количество книг, которые вышли после 1 января 2000 года

In [ ]:

def load_and_print_dataset(query):   
    df = get_dataset(query, connection_string)
    display(df)



In [ ]:
# загрузка и вывод датасета books
ss = '''
           SELECT COUNT(book_id) 
           FROM books
           WHERE publication_date > '2000-01-01'; 
        '''

load_and_print_dataset(ss)

count
0    819

**Вывод:**

- Было выпушено 819 книг, что на 181 меньше от общего кол-ва

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Расчёт верный
</div>

### Количество обзоров и средняя оценка для каждой книги

In [ ]:
dd = '''
    SELECT books.book_id,
            books.title, 
            COUNT(DISTINCT reviews.review_id) AS count_review,  
            ROUND(AVG(ratings.rating),2) AS avg_ratings         
    FROM books
    LEFT JOIN ratings ON ratings.book_id = books.book_id 
    LEFT JOIN reviews ON reviews.book_id = books.book_id 
    GROUP BY books.book_id
    ORDER BY count_review DESC,avg_ratings DESC;
'''   
load_and_print_dataset(dd)

book_id                                              title  count_review  \
0        948                            Twilight (Twilight  #1)             7   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
2        299  Harry Potter and the Chamber of Secrets (Harry...             6   
3        656                                     The Book Thief             6   
4        734                                   The Glass Castle             6   
..       ...                                                ...           ...   
995      191  Disney's Beauty and the Beast (A Little Golden...             0   
996      387                               Leonardo's Notebooks             0   
997      221                          Essential Tales and Poems             0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
999      808                            The Natural Way to Draw             0   

     avg_ratings  
0           3.66  
1           4.41  
2           4.29  
3           4.26  
4           4.21  
..           ...  
995         4.00  
996         4.00  
997         4.00  
998         3.67  
999         3.00  

[1000 rows x 4 columns]

**Вывод:**

- Сумерки и гарри поттер имеет больше всего отзывов и рейтинг

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Грамотный шаг, что добавил колонку с названием книг
</div>

### Издательство, которое выпустило наибольшее число книг толще 50 страниц (исключение из анализа брошюр)

In [ ]:
aa = '''
      SELECT p.publisher, 
          COUNT(b.book_id) AS books_published
      FROM publishers AS p
      INNER JOIN books AS b ON p.publisher_id = b.publisher_id
      WHERE b.num_pages > 50
      GROUP BY p.publisher_id
      ORDER BY books_published DESC
      LIMIT 1;
        '''
load_and_print_dataset(aa)

publisher  books_published
0  Penguin Books               42

**Вывод:**

- Penguin Books	выпустило больше всего книг

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Согласен с результатом. Действительно данное издательство является лидером по выпущенным книгам
</div>

### Автор с самой высокой средней оценкой книг (только книги с 50 и более оценками)

In [ ]:
query = '''
        SELECT a.author, 
                ROUND(AVG(r.rating), 2) AS avg_books_rating
        FROM books AS b
        INNER JOIN authors AS a ON b.author_id = a.author_id
        LEFT OUTER JOIN ratings AS r ON b.book_id = r.book_id
        GROUP BY a.author
        HAVING COUNT(r.rating) > 50
        ORDER BY AVG(r.rating) DESC
        LIMIT 1;
        '''
load_and_print_dataset(query)

author  avg_books_rating
0  J.K. Rowling/Mary GrandPré              4.29

**Вывод:**

- J.K. Rowling/Mary GrandPré имеет средний рейтинг 4.3 - нужно скупить все книги изданные и черновики данного автора

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__
    
После стольких лет? Всегда...

![image.png](attachment:image.png)

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [ ]:
vv = '''
        SELECT ROUND(AVG(review_count)) AS avg_review_count
        FROM (SELECT username,
              COUNT(review_id) AS review_count
              FROM reviews
              WHERE username in (SELECT username
                                 FROM ratings
                                 GROUP BY username
                                 HAVING COUNT(rating_id) > 50)
              GROUP BY username) AS rc;
        '''

load_and_print_dataset(vv)

avg_review_count
0              24.0

**Вывод:**

- Среднее кол-во обзоров от пользователей - 24

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Ты корректно отобрал пользователей и рассчитал их показатели.

### Вывод

**Вывод:**

- Пользователи прояляют хорошую активность
- Необходимо увеличить кол-во книг в базе, обязательно стоит добавить больше книг авторов с хорошим рейтингом
- Penguin Books выпустило больше всего книг
- Среднее кол-во обзоров от пользователей - 24
- J.K. Rowling/Mary GrandPré имеет средний рейтинг 4.3 - нужно скупить все книги изданные и черновики данного автора
- 

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Получилась хорошая работа, развернутый итоговый вывод придает ей вид законченного проекта. Поздравляю с окончанием курса! У тебя получилось справиться со всеми задачами. 
</div>